### ISSUE TITLE
Investigate database deadlock when deleting pipelines on Wings

### ISSUE THREAD

**Issue by chenbh** [concourse/concourse#3383]
For wings, we have a SLI in the form of a [pipeline](https://github.com/concourse/oxygen-mask/blob/595aa529f51aeff77f3ea39992cdc75ba77a3932/monitor.yml#L223-L255) that measures the time it takes to create, run, and delete a new pipeline

It seems to be failing quite frequently, caused by a deadlock in the database
```
Feb 25 08:40:58 0cce4660-e58d-49b8-b945-1e8198c7e40b.web.web.concourse-wings.bosh web/rs2: {"timestamp":"2019-02-25T16:40:57.674000810Z","level":"error","source":"atc","message":"atc.destroying-pipeline.failed","data":{"error":"pq: deadlock detected","name":"new-pipeline","session":"4151005"}}
Feb 25 08:46:06 a5a4055d-b531-46bd-8236-1b8c7e18d579.web.web.concourse-wings.bosh web/rs2: {"timestamp":"2019-02-25T16:46:05.364220543Z","level":"error","source":"atc","message":"atc.destroying-pipeline.failed","data":{"error":"pq: deadlock detected","name":"new-pipeline","session":"5522683"}}
Feb 25 08:49:12 a5a4055d-b531-46bd-8236-1b8c7e18d579.web.web.concourse-wings.bosh web/rs2: {"timestamp":"2019-02-25T16:49:11.793615485Z","level":"error","source":"atc","message":"atc.destroying-pipeline.failed","data":{"error":"pq: deadlock detected","name":"new-pipeline","session":"5527917"}}
Feb 25 08:55:10 0cce4660-e58d-49b8-b945-1e8198c7e40b.web.web.concourse-wings.bosh web/rs2: {"timestamp":"2019-02-25T16:55:10.295787998Z","level":"error","source":"atc","message":"atc.destroying-pipeline.failed","data":{"error":"pq: deadlock detected","name":"new-pipeline","session":"4162964"}}
```

Looking at our Cloud SQL logs:
```
[3739346]: [1-1] db=atc,user=atc ERROR:  deadlock detected 
[3739346]: [2-1] db=atc,user=atc DETAIL:  Process 3739346 waits for ShareLock on transaction 3562502833; blocked by process 3737951. 
	Process 3737951 waits for ShareLock on transaction 3562504310; blocked by process 3739346. 
	Process 3739346: INSERT INTO independent_build_inputs (first_occurrence,input_name,job_id,resource_config_version_id,resource_id) VALUES ($1,$2,$3,$4,$5) 
	Process 3737951: DELETE FROM pipelines WHERE id = $1 
[3739346]: [3-1] db=atc,user=atc HINT:  See server log for query details. 
[3739346]: [4-1] db=atc,user=atc CONTEXT:  while locking tuple (13,13) in relation "jobs" 
```

Other contexts that have been observed:
```
...
[3102934]: [6-1] db=atc,user=atc CONTEXT: while deleting tuple (66,6) in relation "independent_build_inputs"
...
[3816218]: [4-1] db=atc,user=atc CONTEXT: while locking tuple (3,22) in relation "resources"
```

**Comment by pivotal-jwinters** [comment]
Ran an `EXPLAIN ANALYZE` on a delete in wings.

```
Delete on pipelines  (cost=0.27..8.33 rows=1 width=6) (actual time=0.040..0.040 rows=0 loops=1)
  ->  Index Scan using pipelines_pkey on pipelines  (cost=0.27..8.33 rows=1 width=6) (actual time=0.008..0.008 rows=1 loops=1)
        Index Cond: (id = 706465)
Planning time: 0.071 ms
Trigger for constraint resources_pipeline_id_fkey on pipelines: time=0.833 calls=1
Trigger for constraint jobs_pipeline_id_fkey on pipelines: time=0.889 calls=1
Trigger for constraint resource_types_pipeline_id_fkey on pipelines: time=0.681 calls=1
Trigger for constraint builds_pipeline_id_fkey on pipelines: time=0.882 calls=1
Trigger pipeline_build_events_delete_trigger on pipelines: time=5.330 calls=1
Trigger for constraint fkey_resource_id on resources: time=0.684 calls=1
Trigger for constraint resource_disabled_versions_resource_id_fkey on resources: time=0.419 calls=1
Trigger for constraint build_resource_config_version_inputs_resource_id_fkey on resources: time=1955.337 calls=1
Trigger for constraint build_resource_config_version_outputs_resource_id_fkey on resources: time=1653.383 calls=1
Trigger for constraint next_build_inputs_resource_id_fkey on resources: time=5.157 calls=1
Trigger for constraint independent_build_inputs_resource_id_fkey on resources: time=4.386 calls=1
Trigger for constraint resource_config_scopes_resource_id_fkey on resources: time=0.477 calls=1
Trigger for constraint resource_pins_resource_id_fkey on resources: time=0.276 calls=1
Trigger for constraint fkey_job_id on jobs: time=0.456 calls=3
Trigger for constraint fkey_job_id on jobs: time=2.416 calls=3
Trigger for constraint independent_build_inputs_job_id_fkey on jobs: time=0.373 calls=3
Trigger for constraint next_build_inputs_job_id_fkey on jobs: time=0.365 calls=3
Trigger for constraint worker_task_caches_job_id_fkey on jobs: time=0.864 calls=3
Trigger for constraint resource_cache_uses_build_id_fkey on builds: time=0.474 calls=1
Trigger for constraint containers_build_id_fkey on builds: time=3.990 calls=1
Trigger for constraint build_outputs_build_id_fkey on builds: time=0.468 calls=1
Trigger for constraint build_inputs_build_id_fkey on builds: time=0.416 calls=1
Trigger for constraint build_image_resource_caches_build_id_fkey on builds: time=0.483 calls=1
Trigger for constraint jobs_latest_completed_build_id_fkey on builds: time=0.766 calls=1
Trigger for constraint jobs_next_build_id_fkey on builds: time=0.540 calls=1
Trigger for constraint jobs_transition_build_id_fkey on builds: time=0.257 calls=1
Trigger for constraint build_resource_config_version_inputs_build_id_fkey on builds: time=0.142 calls=1
Trigger for constraint build_resource_config_version_outputs_build_id_fkey on builds: time=0.179 calls=1
Trigger for constraint resource_config_versions_resource_config_scope_id_fkey on resource_config_scopes: time=2.479 calls=1
Trigger for constraint resources_resource_config_scope_id_fkey on resource_config_scopes: time=47.827 calls=1
Trigger for constraint next_build_inputs_resource_config_version_id_fkey on resource_config_versions: time=0.268 calls=1
Trigger for constraint independent_build_inputs_resource_config_version_id_fkey on resource_config_versions: time=0.216 calls=1
Execution time: 3693.528 ms
```

This line sticks out:

```
Trigger for constraint resources_resource_config_scope_id_fkey on resource_config_scopes: time=47.827 calls=1
```

**Comment by pivotal-jwinters** [comment]
Our current theory is that this deadlock has always been possible, it was just really unlikely. But something caused the delete to take much longer and as a result, we increase the likelyhood that the deadlock occurs.

**Comment by chenbh** [comment]
Sounds pretty consistent with what the graphs show: the Feb 14th deployment caused a huge jump (from 100ms to 10s) in `http response time` for DeletePipeline
![image](https://user-images.githubusercontent.com/17577073/53512588-1ba55600-3a91-11e9-874a-169714a06e4c.png)

At the same time that's when the SLI started degrading (ignoring the complete outage from the 8th to 11th)
![image](https://user-images.githubusercontent.com/17577073/53511998-a9804180-3a8f-11e9-9423-9e0b50ea5469.png)

**Comment by pivotal-jwinters** [comment]
Looking at the structure of both `build_resource_config_version_inputs` and `build_resource_config_version_outputs` we probably just need to add some indexes.

[A multi column UNIQUE index is only useful as a search index on the first column](https://dba.stackexchange.com/questions/197670/unique-index-on-multiple-columns-vs-single-column-index).

In our case the first column is `build_id` so I think its probably the `resource_id` that needs to be indexed.

\cc @clarafu @vito

**Comment by vito** [comment]
Is this fixed by https://github.com/concourse/concourse/pull/3382?

**Comment by pivotal-jwinters** [comment]
@vito probably?

**Comment by cirocosta** [comment]
A data point from today's deployment:

<img width="1911" alt="screen shot 2019-02-28 at 1 02 28 pm" src="https://user-images.githubusercontent.com/3574444/53587873-786c4380-3b59-11e9-852f-cede362e4def.png">

After the deployment (even though not completely finished yet), seems to get better

**Comment by topherbullock** [comment]
Wings SLI for "Run Existing Job" is performing really poorly due to slow calls to the API for triggering builds: https://snapshot.raintank.io/dashboard/snapshot/zbOkiXli0xpPjyqICCqyXD2wbNf15kL5?panelId=21&fullscreen&orgId=2

Attempting to delete the pipeline whose job we're triggering builds on results in a 500 error : 
```
Mar 07 16:19:33 053691d9-a4f8-4eb8-a2e7-95355095e551.web.web.concourse-wings.bosh web/rs2: {"timestamp":"2019-03-07T21:19:32.490473587Z","level":"error","source":"atc","message":"atc.destroying-pipeline.failed","data":{"error":"pq: deadlock detected","name":"monitoring","session":"4492278"}}
```

It feels like these two issues may be related

**Comment by cirocosta** [comment]
Hey @topherbullock ,

We now have a dimension in such metric that represents status code. Maybe we can leverage that too 🤔 

We're currently using that in `hush-house` through Prometheus like this:

```
concourse_http_responses_duration_seconds_count{
  status=~"5.*",
  kubernetes_pod_name=~"$atc",
  kubernetes_namespace=~"$namespace"
}
```

**Comment by cirocosta** [comment]
Hey,

I was looking at the number of HTTP responses that were delivered by `hush-house`'s web instances with a status code of 500, and it turns out that `DestroyPipeline` returns 500s quite often:

<img width="696" alt="Screen Shot 2019-05-21 at 8 39 35 AM" src="https://user-images.githubusercontent.com/3574444/58096854-00d01280-7ba4-11e9-99b0-0dddbce690b9.png">

Looking at the logs, we can indeed verify that it's indeed related to the deadlocks:

```
{
  "ts": "2019-05-21T12:10:57.998460788Z",
  "msg": "atc.destroying-pipeline.failed",
  "err": "pq: deadlock detected"
}
{
  "ts": "2019-05-21T08:42:17.986094429Z",
  "msg": "atc.destroying-pipeline.failed",
  "err": "pq: deadlock detected"
}
{
  "ts": "2019-05-21T04:58:43.000371227Z",
  "msg": "atc.destroying-pipeline.failed",
  "err": "pq: deadlock detected"
}
{
  "ts": "2019-05-21T03:22:47.979945728Z",
  "msg": "atc.destroying-pipeline.failed",
  "err": "pq: deadlock detected"
}
{
  "ts": "2019-05-20T21:43:43.119048270Z",
  "msg": "atc.destroying-pipeline.failed",
  "err": "pq: deadlock detected"
}
{
  "ts": "2019-05-20T12:50:32.974226620Z",
  "msg": "atc.destroying-pipeline.failed",
  "err": "pq: deadlock detected"
}
```

Those can be indeed seen from the database perspective to:

```
[1316485]: [1-1] db=atc,user=atc ERROR:  deadlock detected 
[1316485]: [2-1] db=atc,user=atc DETAIL:  Process 1316485 waits for ShareLock on transaction 292767174; blocked by process 1323065. 
	Process 1323065 waits for ShareLock on transaction 292767171; blocked by process 1316485. 
	Process 1316485: DELETE FROM pipelines WHERE id = $1 
	Process 1323065: INSERT INTO independent_build_inputs (first_occurrence,input_name,job_id,resource_config_version_id,resource_id) VALUES ($1,$2,$3,$4,$5) 
[1316485]: [3-1] db=atc,user=atc HINT:  See server log for query details. 
[1316485]: [4-1] db=atc,user=atc CONTEXT:  while deleting tuple (5,84) in relation "independent_build_inputs" 
  	SQL statement "DELETE FROM ONLY "public"."independent_build_inputs" WHERE $1 OPERATOR(pg_catalog.=) "job_id"" 
[1316485]: [5-1] db=atc,user=atc STATEMENT:  DELETE FROM pipelines WHERE id = $1 
```
 
(version: 5.2.0)

Thanks!

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
